R/glossary.R

```diff
@@ -0,0 +1,390 @@
+#'
+#' g <- as_glossary(list(ss = "steady state", ALB = "albumin", WT = "weight"))
+#'
+#' glossary_notes(g, ALB, ss)
+#'
+#' @seealso [st_notes_glo()], [read_glossary()]
+#' @export
+glossary_notes <- function(x, ...) UseMethod("glossary_notes")
+
+#' @param ... unquoted glossary labels to be included in the notes text.
+#' @param format passed to [read_glossary()].
+#' @rdname glossary_notes
+#' @export
+glossary_notes.character <- function(x, ..., format = guess_glo_fmt(x)) {
+  x <- read_glossary(x, format = format)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.list <- function(x, format = guess_glo_fmt(x), ...) {
+  x <- as_glossary(x, format = format)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.glossary <- function(x, ..., sep = ": ", collapse = "; ",
+                                    labels = NULL) {
+  labels <- cvec_cs(labels)
+  dots <- eval_select(expr(c(...)), x)
+  labels <- c(names(dots), labels)
+  if(!length(labels)) labels <- names(x)
+  build_glossary_notes(x, labels, sep, collapse)
+}
+
+build_glossary_notes <- function(glossary, labels, sep, collapse) {
+  abort_bad_glo_labels(glossary, labels)
+  glossary <- glossary[labels]
+  abb <- vapply(glossary, "[[", "abbreviation", FUN.VALUE = "a")
+  notes <- lapply(glossary, "[[", "definition")
+  notes <- paste0(abb, sep, notes)
+  if(is.character(collapse)) {
+    notes <- paste0(notes, collapse = collapse)
+  }
+  notes
+}
+
+#' @export
+print.glossary_entry <- function(x, ...) {
+  cat(ans <- paste0(x$definition, " (", x$abbreviation, ")"))
+}
+
+#' @export
+print.glossary <- function(x, ...) {
+  label <- names(x)
+  def <- map_chr(x, "definition")
+  def <- substr(def, 0, 40)
+  pad <- ifelse(nchar(def) >= 40, "...", "")
+  def <- paste0(def, pad)
+  label <- formatC(label, width = max(nchar(label)), flag = "-")
+  cat(paste0(label, " : ", def), sep = "\n")
+}
+
+#' @export
+as.data.frame.glossary <- function(x, ...) {
+  data.frame(
+    label = names(x),
+    definition = map_chr(x, "definition"),
+    abbreviation = map_chr(x, "abbreviation"),
+    row.names = NULL
+  )
+}
+
+#' @export
+as.list.glossary <- function(x, ...) {
+  class(x) <- "list"
+  x <- lapply(x, as.list)
+  x
+}
+
+#' @export
+as.list.glossary_entry <- function(x, ...) {
+  class(x) <- "list"
+  x
+}
+
+#' @export
+c.glossary <- function(x, y, ...) {
+  cl <- class(x)
+  new_cols <- setdiff(names(y), names(x))
+  if(!identical(new_cols, names(y))) {
+    stop("`x` and `y` cannot share any names.")
+  }
+  ans <- structure(c(unclass(x), unclass(y)), class = cl)
+  more <- list(...)
+  if(length(more)) {
+    return(c(ans, more[[1]]))
+  }
+  ans
+}
+
+#' @export
+`$.glossary` <- function(x, name, ...) {
+  if(!exists(name, x)) {
+    return(NULL)
+  }
+  unclass(x)[[name]]
+}
+
+#' @export
+`[.glossary` <- function(x, i, j, drop = FALSE) {
+  cl <- class(x)
+  x <- unclass(x)
+  x <- x[i]
+  structure(x, class = cl)
+}
```

Comment:
Could this be simplified to this?

```R
#' @export
`[.glossary` <- function(x, ...) {
  structure(NextMethod(), class = class(x))
}
```

Reply:
Thanks; much simpler and I think it retains the class attribute of the pre-subset object which is what set a lot of this off.